### PR TITLE
Emits metrics to monitor divergence of shards as db repair progresses

### PR DESCRIPTION
**Pull request by rishabhjain08**
**What this PR does / why we need it**:

This PR emits metrics from db repair process which represents the maximum divergence of a block from any of its peers. These metrics can be used to monitor -
* Divergence does not increase over time time.
* Maximum amount of data diverged between a origin block and a peer block.
* Maximum amount of data diverged as percentage of origin block's size.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/dbnode/storage/repair.go [hunk 1]
```diff
@@ -274,15 +282,15 @@ func (r shardRepairer) Repair(
 		session := sessions[i].session
 		perSeriesReplicaIter, err := session.FetchBlocksFromPeers(nsMeta, shard.ID(), level, metadatasToFetchBlocksFor, rsOpts)
 		if err != nil {
-			return repair.MetadataComparisonResult{}, err
+			return metadataRes, err
```
Comment (justinjc) [inline src/dbnode/storage/repair.go:285]: Why the change here? It's a pretty common Go pattern to return nil(-equivalent) values when returning an error.

File: src/dbnode/storage/repair.go [hunk 2]
```diff
@@ -206,6 +210,10 @@ func (r shardRepairer) Repair(
 		seriesWithChecksumMismatches        = metadataRes.ChecksumDifferences.Series()
 	)
 
+	// Shard repair can fail due to transient network errors due to the significant amount of data fetched from peers.
+	// So collect and emit metadata comparision metrics before fetching blocks from peer to repair.
```
[the remark this answers is not in the document]
Comment (justinjc) [reply src/dbnode/storage/repair.go:214]: Cool, this makes sense. Good to have these stats even though the repair fails.

**Comment by justinjc** [comment]
This diff looks good to me in general with a few minor things. Looking at it though, it seems like computing these metrics are rather expensive, as we're looping through all mismatched blocks for all mismatched series and doing some computation, which could be in the many thousands.

How often do repairs happen? I suppose if it's not very frequent, looping through everything is okay. What's the objective of these metrics? i.e. are there new actionable things we can do with these that we aren't able to with the repair metrics available today?

File: src/dbnode/storage/repair_test.go [hunk 3]
```diff
@@ -558,6 +561,21 @@ func TestDatabaseShardRepairerRepairMultiSession(t *testing.T) {
 		{Host: hosts[1], Metadata: inputBlocks[1].Metadata},
 	}
 	require.Equal(t, expected, currBlock.Metadata())
+
+	// Validate the expected metrics were emitted
+	scopeSnapshot := scope.Snapshot()
+	countersSnapshot := scopeSnapshot.Counters()
+	gaugesSnapshot := scopeSnapshot.Gauges()
+	require.Equal(t, int64(2),
+		countersSnapshot["repair.series+namespace=testNamespace,resultType=total,shard=0"].Value())
```
Comment (andrewmains12) [inline src/dbnode/storage/repair_test.go:570]: There's a helper you can use to generate this key btw: `tally.KeyForPrefixedStringMap`

File: src/dbnode/storage/repair_test.go [hunk 4]
```diff
@@ -499,29 +503,16 @@ func TestDatabaseShardRepairerRepairMultiSession(t *testing.T) {
...
-		resDiff = diffRes
-	}
 
 	var (
 		ctx   = context.NewContext()
 		nsCtx = namespace.Context{ID: namespaceID}
 	)
 	require.NoError(t, err)
-	repairer.Repair(ctx, nsCtx, nsMeta, repairTimeRange, shard)
+	resDiff, _ := repairer.Repair(ctx, nsCtx, nsMeta, repairTimeRange, shard)
```
[the remark this answers is not in the document]
Comment (andrewmains12) [reply src/dbnode/storage/repair_test.go:514]: +1 yeah--just `require.NoError` it

File: src/dbnode/storage/repair.go [hunk 5]
```diff
@@ -56,7 +56,11 @@ var (
 	errRepairInProgress = errors.New("repair already in progress")
 )
 
-type recordFn func(namespace ident.ID, shard databaseShard, diffRes repair.MetadataComparisonResult)
+type recordFn func(
+	origin topology.Host,
+	namespace ident.ID,
+	shard databaseShard,
+	diffRes repair.MetadataComparisonResult)
```
[the remark this answers is not in the document]
Comment (rishabhjain08) [reply src/dbnode/storage/repair.go:63]: Done.

File: src/dbnode/storage/repair.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rishabhjain08) [reply src/dbnode/storage/repair.go:214]: Fixed.

File: src/dbnode/storage/repair.go [hunk 6]
```diff
@@ -377,11 +384,65 @@ func (r shardRepairer) recordDifferences(
...
+
 	// Record checksum differences.
 	checksumDiffScope.Counter("series").Inc(diffRes.ChecksumDifferences.NumSeries())
 	checksumDiffScope.Counter("blocks").Inc(diffRes.ChecksumDifferences.NumBlocks())
 }
 
+// computeMaximumBlockSizeDifferenceAsPercentage returns a metric which represents maximum divergence of a shard with
+// any of its peers. A positive divergence means that origin shard has more data than its peer and a negative
+// divergence means that origin shard has lesser data than its peer.  Since sizes for all the blocks in rentention
+// window are not redily available, exact divergence of a shard from its peer cannot be calculated. So this method
```
[the remark this answers is not in the document]
Comment (rishabhjain08) [reply src/dbnode/storage/repair.go:399]: Fixed.

File: src/dbnode/storage/repair.go [hunk 7]
```diff
@@ -377,11 +384,65 @@ func (r shardRepairer) recordDifferences(
...
+// settles for returning maximum divergence of a block/shard with any of its peers. Divergence(as percentage) of shard
+// is upper bounded by divergence of block/shard so this metric can be used to monitor severity of divergence.
+func (r shardRepairer) computeMaximumBlockSizeDifference(
+	origin topology.Host,
+	diffRes repair.MetadataComparisonResult,
+) (int64, float64) {
+	var maxBlockSizeDiffAsRatio float64 = 0
```
[the remark this answers is not in the document]
Comment (rishabhjain08) [reply src/dbnode/storage/repair.go:406]: Fixed.

File: src/dbnode/storage/repair_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rishabhjain08) [reply src/dbnode/storage/repair_test.go:514]: Fixed.

File: src/dbnode/storage/repair.go [hunk 8]
```diff
@@ -377,11 +384,65 @@ func (r shardRepairer) recordDifferences(
...
+	// Iterate over all the series which differ in size between origin and a peer.
+	for _, entry := range diffRes.SizeDifferences.Series().Iter() {
+		series := entry.Value()
+		replicaBlocksMetadata := diffRes.SizeDifferences.GetOrAdd(series.ID)
+		// Iterate over all the time ranges which had a mismatched series between origin and a peer.
+		for _, replicasMetadata := range replicaBlocksMetadata.Blocks() {
+			// Setting minimum origin block size to 1 so that percetages off of origin block size can be calculated
+			// without worrying about divide by zero errors. Exact percentages are not required so setting a non-zero
+			// size for an empty block is acceptable.
+			var originBlockSize int64 = 1
```
[the remark this answers is not in the document]
Comment (rishabhjain08) [reply src/dbnode/storage/repair.go:417]: Fixed.

File: src/dbnode/storage/repair.go [hunk 9]
```diff
@@ -377,11 +384,65 @@ func (r shardRepairer) recordDifferences(
...
+			// Represents maximum size difference of a block with one of its peers.
+			var maxPeerBlockSizeDiff int64 = 0
+			// Record the block size on the origin.
+			for _, replicaMetadata := range replicasMetadata.Metadata() {
+				if replicaMetadata.Host.ID() == origin.ID() {
+					originBlockSize = int64(math.Max(float64(originBlockSize), float64(replicaMetadata.Size)))
```
[the remark this answers is not in the document]
Comment (rishabhjain08) [reply src/dbnode/storage/repair.go:423]: Fixed.

File: src/dbnode/storage/repair.go [hunk 10]
```diff
@@ -377,11 +384,65 @@ func (r shardRepairer) recordDifferences(
...
+					break
+				}
+			}
+			// Fetch the maximum block size difference of origin with any of its peers.
+			for _, replicaMetadata := range replicasMetadata.Metadata() {
+				if replicaMetadata.Host.ID() != origin.ID() {
+					blockSizeDiff := originBlockSize - replicaMetadata.Size
+					if math.Abs(float64(blockSizeDiff)) > math.Abs(float64(maxPeerBlockSizeDiff)) {
```
[the remark this answers is not in the document]
Comment (rishabhjain08) [reply src/dbnode/storage/repair.go:431]: I couldnt figure out a better way to do this. IfI change sizeDiffs variables to float64 I will have to use int64 casting instead of float64 casting.

File: src/dbnode/storage/repair.go (hunk 1 above)
Comment (rishabhjain08) [reply src/dbnode/storage/repair.go:285]: I am changing it back to nil. Although the thought process behind this change was that we are throwing away work done to compare so its better to return with an error and leave it to the caller to decide if they want to log result even in case of an error.

File: src/dbnode/storage/repair.go [hunk 11]
```diff
@@ -377,11 +384,65 @@ func (r shardRepairer) recordDifferences(
...
+	sizeDiffScope.Gauge("max-block-size-diff-as-percentage").Update(blockSizeDiffAsPercentage)
+
 	// Record checksum differences.
 	checksumDiffScope.Counter("series").Inc(diffRes.ChecksumDifferences.NumSeries())
 	checksumDiffScope.Counter("blocks").Inc(diffRes.ChecksumDifferences.NumBlocks())
 }
 
+// computeMaximumBlockSizeDifferenceAsPercentage returns a metric which represents maximum divergence of a shard with
+// any of its peers. A positive divergence means that origin shard has more data than its peer and a negative
+// divergence means that origin shard has lesser data than its peer.  Since sizes for all the blocks in rentention
```
[the remark this answers is not in the document]
Comment (rishabhjain08) [reply src/dbnode/storage/repair.go:398]: On graph we will have to filter metrics by cluster. One we are looking at metrics for the cluster, all the nodes in that cluster become our origin nodes.

File: src/dbnode/storage/repair_test.go (hunk 3 above)
Comment (rishabhjain08) [reply src/dbnode/storage/repair_test.go:570]: Fixed.

**Comment by rishabhjain08** [comment]
> This diff looks good to me in general with a few minor things. Looking at it though, it seems like computing these metrics are rather expensive, as we're looping through all mismatched blocks for all mismatched series and doing some computation, which could be in the many thousands.
> 
> How often do repairs happen? I suppose if it's not very frequent, looping through everything is okay. What's the objective of these metrics? i.e. are there new actionable things we can do with these that we aren't able to with the repair metrics available today?



> This diff looks good to me in general with a few minor things. Looking at it though, it seems like computing these metrics are rather expensive, as we're looping through all mismatched blocks for all mismatched series and doing some computation, which could be in the many thousands.
> 
> How often do repairs happen? I suppose if it's not very frequent, looping through everything is okay. What's the objective of these metrics? i.e. are there new actionable things we can do with these that we aren't able to with the repair metrics available today?


1. Repairs happens once every [40 minutes in staging](https://github.com/m3db/m3/blob/4154ff1894f003ae5eef05e77358f99a714af7fc/src/cmd/services/m3dbnode/config/config.go#L325). 
2. One repair run only repairs one block per shard. So we will only loop exactly once in in this function introduced by this PR.
3. All the computation is done from in-memory objects so it should complete in microseconds even for thousands of blocks. As such I don't think this function can cause performance issues given the low frequency with with repair workflow runs.
